zarrtraj/ZARR.py

```diff
@@ -0,0 +1,1567 @@
+            )
+
+    def __iter__(self):
+        """Iterate over all frames in the trajectory"""
+        self._reopen()
+        self._frame_seq = collections.deque(range(0, self.n_frames))
+        self._cache.update_frame_seq(self._frame_seq)
+        return self
+
+    def next(self):
+        if self._frame_seq is None and self._frame + 1 < self.n_frames:
+            self._frame_seq = collections.deque([self._frame + 1])
+            self._cache.update_frame_seq(self._frame_seq)
+        elif self._frame_seq is None:
+            self.rewind()
+            raise StopIteration from None
+        try:
+            ts = self._read_next_timestep()
+        except (EOFError, IOError):
+            self.rewind()
+            raise StopIteration from None
+        else:
+            for auxname, reader in self._auxs.items():
+                ts = self._auxs[auxname].update_ts(ts)
+
+            ts = self._apply_transformations(ts)
+
+        return ts
+
+    def iter_as_aux(self, auxname):
+        """Iterate over the trajectory with an auxiliary reader"""
+        aux = self._check_for_aux(auxname)
+        self._reopen()
+        self._frame_seq = collections.deque(range(0, self.n_frames))
+        self._cache.update_frame_seq(self._frame_seq)
+        aux._restart()
+        while True:
+            try:
+                yield self.next_as_aux(auxname)
+            except StopIteration:
+                return
+
+    def copy(self):
+        """Return independent copy of this Reader.
+
+        New Reader will have its own file handle and can seek/iterate
+        independently of the original.
+
+        Will also copy the current state of the Timestep held in the original
+        Reader.
+
+
+        .. versionchanged:: 2.2.0
+           Arguments used to construct the reader are correctly captured and
+           passed to the creation of the new class. Previously the only
+           ``n_atoms`` was passed to class copies, leading to a class created
+           with default parameters which may differ from the original class.
+        """
+
+        new = self.__class__(**self._kwargs)
+
+        if self.transformations:
+            new.add_transformations(*self.transformations)
+        # seek the new reader to the same frame we started with
+        new[self.ts.frame]
+        # then copy over the current Timestep in case it has
+        # been modified since initial load
+        new.ts = self.ts.copy()
+        new._cache._timestep = new.ts
+        for auxname, auxread in self._auxs.items():
+            new.add_auxiliary(auxname, auxread.copy())
+        return new
+
+    @property
+    def n_frames(self):
+        """number of frames in trajectory"""
+        return len(self._global_steparray)
+
+    @staticmethod
+    def _format_hint(thing):
+        """Can this Reader read *thing*"""
+        # nb, filename strings can still get passed through if
+        # format='H5MD' is used
+        pass
```

Comment:
I  assume we need this?

Reply:
Yes, this is actually something I want your thoughts on- should this reader take precedence over the core H5MDReader for reading H5MD files from disk if the user imports it? If so, I can just catch all files with the .zarrmd or .h5md file extension and don't need a format hint. If not, when and how should the user specify they want to use this reader to read their h5md file?